### PR TITLE
Return remaining bytes from exchange sources

### PR DESCRIPTION
**Pull request by Yuhta**
Summary:
This is the first diff to upgrade the exchange protocol.  This change
only exposes the remaining bytes to buffer manager; it does not change the
existing protocol yet, and is compatible with the current Prestissimo code.

Also added a few statistics to spot skewed exchange.

See https://github.com/prestodb/presto/issues/21926 for details about the
design.

Differential Revision: D53793123

### CODE REVIEW

File: velox/exec/OutputBuffer.h [hunk 1]
```diff
@@ -78,6 +81,9 @@ class ArbitraryBuffer {
   /// there are sufficient buffered pages.
   std::vector<std::shared_ptr<SerializedPage>> getPages(uint64_t maxBytes);
 
+  /// Append the available page sizes to `out'.
+  void getAvailablePageSizes(std::vector<int64_t>& out) const;
```
Comment (mbasmanova) [inline velox/exec/OutputBuffer.h:85]: Any particular reason not to return std::vector?

File: velox/exec/OutputBuffer.h [hunk 2]
```diff
@@ -125,13 +131,24 @@ class DestinationBuffer {
   /// arbitrary buffer on demand.
   void loadData(ArbitraryBuffer* buffer, uint64_t maxBytes);
 
+  struct Data {
+    std::vector<std::unique_ptr<folly::IOBuf>> data;
+    std::vector<int64_t> remainingBytes;
+    bool immediate;
```
Comment (mbasmanova) [inline velox/exec/OutputBuffer.h:137]: nit: it would be nice to document this struct

File: velox/exec/OutputBuffer.h [hunk 3]
```diff
@@ -125,13 +131,24 @@ class DestinationBuffer {
...
+
   /// Returns a shallow copy (folly::IOBuf::clone) of the data starting at
   /// 'sequence', stopping after exceeding 'maxBytes'. If there is no data,
   /// 'notify' is installed so that this gets called when data is added. If not
   /// null, 'activeCheck' is used to check if the consumer of a destination
   /// buffer with 'notify' installed is currently active or not. This only
   /// applies for arbitrary output buffer for now.
-  std::vector<std::unique_ptr<folly::IOBuf>> getData(
+  ///
+  /// When arbitraryBuffer is provided, and the this buffer is not at end (no
```
Comment (mbasmanova) [inline velox/exec/OutputBuffer.h:147]: typos in "and the this buffer"

File: velox/exec/OutputBuffer.h [hunk 4]
```diff
@@ -223,6 +242,9 @@ class OutputBuffer {
     /// Average time each piece of data has been buffered for in milliseconds.
     int64_t averageBufferTimeMs{0};
 
+    /// The number of largest buffers that handle 80% of the total data.
+    int32_t numTopBuffers{0};
```
Comment (mbasmanova) [inline velox/exec/OutputBuffer.h:246]: nit: it would be nice to extract adding new stats into a separate PR

**Review by mbasmanova — APPROVED** [review]
Thanks.

File: velox/exec/OutputBuffer.h (hunk 1 above)
Comment (Yuhta) [reply velox/exec/OutputBuffer.h:85]: So we don't need to create a new vector every time and just appending the existing one already created for DestinationBuffer
